File: engine/models/contract.py
from __future__ import annotations
from dataclasses import dataclass, field

@dataclass
class ContractIncentive:
    description: str
    criteria: str
    amount: int
    incentive_type: str

@dataclass
class ContractYear:
    year: int
    salary: int
    is_guaranteed: bool = True
    guarantee_date: str | None = None
    incentives: list[ContractIncentive] = field(default_factory=list)
    trade_bonus: int = 0

@dataclass
class TradeRestriction:
    restriction_type: str
    restriction_end_date: str

@dataclass
class ContractInfo:
    id: str
    player_id: str
    team_id: str
    contract_type: str
    years: list[ContractYear] = field(default_factory=list)
    total_value: int = 0
    signing_date: str = ""
    bird_rights_status: str = "none"
    trade_restriction: TradeRestriction | None = None
    has_no_trade_clause: bool = False
    has_player_option: bool = False
    has_team_option: bool = False
    player_option_year: int | None = None
    team_option_year: int | None = None
    is_fully_guaranteed: bool = False
    poison_pill_provision: bool = False

@dataclass
class CBAConstants:
    salary_cap: int
    luxury_tax_threshold: int
    first_apron: int
    second_apron: int
    minimum_team_salary: int
    max_contract_percentages: dict = field(default_factory=dict)
    rookie_scale: dict = field(default_factory=dict)
    veteran_minimums: dict = field(default_factory=dict)
    mid_level_exception: int = 0
    taxpayer_mle: int = 0
    bi_annual_exception: int = 0
    annual_raises: dict = field(default_factory=dict)
    trade_rules: dict = field(default_factory=dict)
    hard_cap_triggers: list[str] = field(default_factory=list)

@dataclass
class RookieScaleEntry:
    year1: int
    year2: int
    year3_option: int
    year4_option: int

@dataclass
class TradeTeamPackage:
    team_id: str
    players_out: list[str] = field(default_factory=list)
    players_in: list[str] = field(default_factory=list)
    picks_out: list = field(default_factory=list)
    picks_in: list = field(default_factory=list)
    cash_out: int = 0
    cash_in: int = 0
    salary_out: int = 0
    salary_in: int = 0

@dataclass
class TradePackage:
    id: str
    teams: list[TradeTeamPackage] = field(default_factory=list)
    status: str = "proposed"
    validation_errors: list[str] = field(default_factory=list)

File: frontend/src/types/contract.ts
import type { DraftPickAsset } from './team';

export interface ContractInfo {
  id: string;
  playerId: string;
  teamId: string;
  type: ContractType;
  years: ContractYear[];
  totalValue: number;
  signingDate: string;
  birdRightsStatus: BirdRightsStatus;
  tradeRestriction: TradeRestriction | null;
  hasNoTradeClause: boolean;
  hasPlayerOption: boolean;
  hasTeamOption: boolean;
  playerOptionYear: number | null;
  teamOptionYear: number | null;
  isFullyGuaranteed: boolean;
  poisonPillProvision: boolean;
}

export interface ContractYear {
  year: number;
  salary: number;
  isGuaranteed: boolean;
  guaranteeDate: string | null;
  incentives: ContractIncentive[];
  tradeBonus: number;
}

export interface ContractIncentive {
  description: string;
  amount: number;
  type: "likely" | "unlikely";
  criteria: string;
}

export type ContractType =
  | "rookie_scale"
  | "rookie_scale_extension"
  | "veteran_max"
  | "designated_veteran_max"
  | "standard"
  | "minimum"
  | "mid_level_exception"
  | "bi_annual_exception"
  | "taxpayer_mid_level"
  | "two_way"
  | "ten_day"
  | "sign_and_trade"
  | "qualifying_offer";

export type BirdRightsStatus =
  | "full_bird"
  | "early_bird"
  | "non_bird"
  | "none";

export interface TradeRestriction {
  type: "newly_signed" | "sign_and_trade" | "poison_pill";
  restrictionEndDate: string;
}

export interface CBAConstants {
  salaryCap: number;
  luxuryTaxThreshold: number;
  firstApron: number;
  secondApron: number;
  minimumTeamSalary: number;
  maxContractPercentages: {
    zeroToSixYears: number;
    sevenToNineYears: number;
    tenPlusYears: number;
  };
  rookieScale: Record<number, RookieScaleEntry>;
  veteranMinimums: Record<number, number>;
  midLevelException: number;
  taxpayerMLE: number;
  biAnnualException: number;
  annualRaises: {
    birdRights: number;
    otherTeam: number;
  };
  tradeRules: {
    overCapMatchingPercentage: number;
    overCapMatchingFlat: number;
    apronMatchingRule: "dollar_for_dollar" | "standard";
  };
  hardCapTriggers: string[];
}

export interface RookieScaleEntry {
  year1: number;
  year2: number;
  year3Option: number;
  year4Option: number;
}

export interface TradePackage {
  id: string;
  teams: TradeTeamPackage[];
  status: "proposed" | "accepted" | "rejected" | "invalid";
  validationErrors: string[];
}

export interface TradeTeamPackage {
  teamId: string;
  playersOut: string[];
  playersIn: string[];
  picksOut: DraftPickAsset[];
  picksIn: DraftPickAsset[];
  cashOut: number;
  cashIn: number;
  salaryOut: number;
  salaryIn: number;
}

export interface Trade {
  id: string;
  date: string;
  teams: TradeTeamPackage[];
  description: string;
  seasonYear: number;
}

export interface CapSheet {
  salaryCap: number;
  totalPayroll: number;
  luxuryTaxThreshold: number;
  firstApron: number;
  secondApron: number;
  capSpace: number;
  midLevelException: number;
  biAnnualException: number;
  taxpayerMLE: number;
  isOverCap: boolean;
  isInLuxuryTax: boolean;
}
